consumer.go

```diff
@@ -240,6 +240,7 @@ func (c *Consumer) mainLoop() {
 		var notification *Notification
 		if c.client.config.Group.Return.Notifications {
 			notification = newNotification(c.subs.Info())
+			c.handleNotification(notification)
```

Comment:
sorry, I don't quite understand your comment. could you please elaborate?

Reply:
So assume there's a loop with a switch to read messages, notifications, and errors. So at some point caller receives a rebalance start notification, and it will synchronise to external storage, mark offsets and continue looping. Now there's a race between the time this notification was emitted and the time consumption from assignments is stopped. If the caller is faster, it will manage to read messages in this time window, but will not mark offsets because a) it doesn't know rebalance is in progress b) batch isn't complete. So these messages will be read again by another consumer after rebalance, and written to external storage twice effectively. Obviously the solution is to put logic after receiving `RebalanceStart` notification to stop reading messages and wait for either `Rebalance{Ok,Error}`, so it'c clearly a caller mistake. But it's not obvious and could be simply avoided by emitting the notification later, after the topic/partition consumers are stopped.